spark/src/main/java/org/apache/iceberg/spark/source/StreamingOffset.java

```diff
@@ -0,0 +1,120 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.io.StringWriter;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.util.JsonUtil;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+
+public class StreamingOffset extends Offset {
+  private static final int CURR_VERSION = 1;
+
+  private static final String VERSION = "version";
+  private static final String SNAPSHOT_ID = "snapshot_id";
+  private static final String INDEX = "index";
+  private static final String IS_STARTING = "is_starting";
+
+  private final long snapshotId;
+  private final int index;
+  private final boolean isStarting;
+
+  public StreamingOffset(long snapshotId, int index, boolean isStarting) {
+    this.snapshotId = snapshotId;
+    this.index = index;
+    this.isStarting = isStarting;
+  }
+
+  public static StreamingOffset fromJson(String json) {
+    Preconditions.checkNotNull(json, "The input JSON string is null");
+
+    try {
+      JsonNode node = JsonUtil.mapper().readValue(json, JsonNode.class);
+      int version = JsonUtil.getInt(VERSION, node);
+      if (version != CURR_VERSION) {
+        throw new IOException("The version number in JSON string " + version + " is not a valid version number");
+      }
+
+      long snapshotId = JsonUtil.getLong(SNAPSHOT_ID, node);
+      int index = JsonUtil.getInt(INDEX, node);
+      boolean isStarting = JsonUtil.getBool(IS_STARTING, node);
+
+      return new StreamingOffset(snapshotId, index, isStarting);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to parse StreamingOffset from JSON string %s", json);
+    }
+  }
+
+  @Override
+  public String json() {
+    StringWriter writer = new StringWriter();
+    try {
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      generator.writeStartObject();
+      generator.writeNumberField(VERSION, CURR_VERSION);
+      generator.writeNumberField(SNAPSHOT_ID, snapshotId);
+      generator.writeNumberField(INDEX, index);
+      generator.writeBooleanField(IS_STARTING, isStarting);
+      generator.writeEndObject();
+      generator.flush();
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to write StreamingOffset to json");
+    }
+
+    return writer.toString();
+  }
+
+  public long snapshotId() {
```

Comment:
It would be nice to have Javadoc for these methods. I haven't read much of the implementation yet, so it isn't clear what `index` is indexing.

Reply:
Sure, I will add the docs explaining them.

Reply:
Fixed.